File: ch8/turtle_controller_cpp/src/turtle_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"
#include "turtlesim/srv/set_pen.hpp"
#include "my_robot_interfaces/srv/activate_turtle.hpp"

using SetPen = turtlesim::srv::SetPen;
using ActivateTurtle = my_robot_interfaces::srv::ActivateTurtle;
using namespace std::chrono_literals;
using namespace std::placeholders;


class TurtleControllerNode : public rclcpp::Node
{
public:
    TurtleControllerNode() : Node("turtle_controller"), is_active_(true), previous_x_(0.0)
    {
        this->declare_parameter("color_1", std::vector<int64_t>{255, 0, 0});
        this->declare_parameter("color_2", std::vector<int64_t>{0, 255, 0});
        this->declare_parameter("turtle_velocity", 1.0);
        color_1_ = this->get_parameter("color_1").as_integer_array();
        color_2_ = this->get_parameter("color_2").as_integer_array();
        turtle_velocity_ = this->get_parameter("turtle_velocity").as_double();


        set_pen_client_ = this->create_client<SetPen>("/turtle1/set_pen");
        activate_turtle_service_ = this->create_service<ActivateTurtle>(
            "activate_turtle", std::bind(&TurtleControllerNode::callbackActivateTurtle, this, _1, _2));
        cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
            "/turtle1/cmd_vel", 10);
        pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtle1/pose", 10, std::bind(&TurtleControllerNode::poseCallback, this, _1));
    }

    void poseCallback(const turtlesim::msg::Pose::SharedPtr pose)
    {
        if (is_active_) {
            auto cmd = geometry_msgs::msg::Twist();
            if (pose->x < 5.5) {
                cmd.linear.x = turtle_velocity_;
                cmd.angular.z = turtle_velocity_;
            }
            else {
                cmd.linear.x = turtle_velocity_* 2.0;
                cmd.angular.z = turtle_velocity_* 2.0;
            }
            cmd_vel_pub_->publish(cmd);

            if (pose->x > 5.5 && previous_x_ <= 5.5) {
                previous_x_ = pose->x;
                RCLCPP_INFO(this->get_logger(), "Set color 1.");
                callSetPen(color_1_.at(0), color_1_.at(1), color_1_.at(2));
            }
            else if (pose->x <= 5.5 && previous_x_ > 5.5) {
                previous_x_ = pose->x;
                RCLCPP_INFO(this->get_logger(), "Set color 2.");
                callSetPen(color_2_.at(0), color_2_.at(1), color_2_.at(2));
            }
        }
    }

    void callSetPen(int r, int g, int b)
    {
        while (!set_pen_client_->wait_for_service(1s)) {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server...");
        }
        auto request = std::make_shared<SetPen::Request>();
        request->r = r;
        request->g = g;
        request->b = b;
        set_pen_client_->async_send_request(
            request, std::bind(&TurtleControllerNode::callbackSetPenResponse, this, _1));
    }

private:

    void callbackSetPenResponse(rclcpp::Client<SetPen>::SharedFuture future)
    {
        (void)future;
        RCLCPP_INFO(this->get_logger(), "Successfully changed pen color");
    }

    void callbackActivateTurtle(const ActivateTurtle::Request::SharedPtr request, 
                              const ActivateTurtle::Response::SharedPtr response)
    {
        is_active_ = request->activate;
        if (request->activate) {
            response->message = "Starting the turtle";
        }
        else {
            response->message = "Stopping the turtle";
        }
    }

    std::vector<int64_t> color_1_;
    std::vector<int64_t> color_2_;
    double turtle_velocity_;
    bool is_active_;
    double previous_x_;
    rclcpp::Client<SetPen>::SharedPtr set_pen_client_;
    rclcpp::Service<ActivateTurtle>::SharedPtr activate_turtle_service_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_sub_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleControllerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ch9/turtle_controller/turtle_controller/turtle_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
from turtlesim.srv import SetPen
from my_robot_interfaces.srv import ActivateTurtle


class TurtleControllerNode(Node):
    def __init__(self):
        super().__init__("turtle_controller")
        self.declare_parameter("color_1", [255, 0, 0])
        self.declare_parameter("color_2", [0, 255, 0])
        self.declare_parameter("turtle_velocity", 1.0)
        self.color_1_ = self.get_parameter("color_1").value
        self.color_2_ = self.get_parameter("color_2").value
        self.turtle_velocity_ = self.get_parameter("turtle_velocity").value

        self.is_active_ = True
        self.previous_x_ = 0.0
        self.set_pen_client_ = self.create_client(SetPen, "/turtle1/set_pen")
        self.activate_turtle_service_ = self.create_service(ActivateTurtle, "activate_turtle", self.callback_activate_turtle)
        self.cmd_vel_pub_ = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        self.pose_sub_ = self.create_subscription(Pose, "/turtle1/pose", self.callback_pose, 10)

    def callback_pose(self, pose: Pose):
        if self.is_active_:
            cmd = Twist()
            if pose.x < 5.5:
                cmd.linear.x = self.turtle_velocity_
                cmd.angular.z = self.turtle_velocity_
            else:
                cmd.linear.x = self.turtle_velocity_ * 2.0
                cmd.angular.z = self.turtle_velocity_ * 2.0
            self.cmd_vel_pub_.publish(cmd)

            if pose.x > 5.5 and self.previous_x_ <= 5.5:
                self.previous_x_ = pose.x
                self.get_logger().info("Set color 1.")
                self.call_set_pen(self.color_1_[0], self.color_1_[1], self.color_1_[2])
            elif pose.x <= 5.5 and self.previous_x_ > 5.5:
                self.previous_x_ = pose.x
                self.get_logger().info("Set color 2.")
                self.call_set_pen(self.color_2_[0], self.color_2_[1], self.color_2_[2])

    def call_set_pen(self, r, g, b):
        while not self.set_pen_client_.wait_for_service(1.0):
            self.get_logger().warn("Waiting for service...")
        request = SetPen.Request()
        request.r = r
        request.g = g
        request.b = b
        future = self.set_pen_client_.call_async(request)
        future.add_done_callback(self.callback_set_pen_response)

    def callback_set_pen_response(self, future):
        self.get_logger().info("Successfully changed pen color")

    def callback_activate_turtle(self, request: ActivateTurtle.Request, response: ActivateTurtle.Response):
        self.is_active_ = request.activate
        if request.activate:
            response.message = "Starting the turtle"
        else:
            response.message = "Stopping the turtle"
        return response


def main(args=None):
    rclpy.init(args=args)
    node = TurtleControllerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()